### DISCUSSION TITLE
LinAlgError: Matrix is not positive definite

### DISCUSSION THREAD

**Discussion by liangsihan** [simnibs/simnibs#222] [7.Creating head models]
Hi, I encountered a error LinAlgError: Matrix is not positive definite when creating a head model using charm. May I ask what might be the problem here? Below is the log

**Comment by liangsihan** [comment]
INFO: simnibs version 4.0.1
INFO: charm run started: Tue Oct 31 01:43:11 2023
DEBUG: options: bayi bayi_T1w.nii.gz --forceqform
INFO: Reading settings from /Users/sihanliang/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/simnibs/charm.ini
DEBUG: {'preprocess': {'denoise': False}, 'samseg': {'threads': 8, 'atlas_name': 'charm_atlas_mni', 'gmm_parameter_file': '', 'init_type': 'atlas'}, 'initmni': {'translation_scale': -100, 'max_iter': 300, 'shrink_factors': [2, 1, 0], 'bg_value': 0, 'smoothing_factors': [4.0, 2.0, 0.0], 'center_of_mass': True, 'samp_factor': 1.0}, 'initatlas': {'affine_scales': [[0.85, 0.85, 0.85], [0.9, 0.9, 0.85], [0.95, 0.95, 0.85]], 'affine_rotations': [-7, -3.5, 0, 3.5, 7], 'affine_horizontal_shifts': [-20.0, -10.0, 0, 10.0, 20.0], 'affine_vertical_shifts': [-10.0, 0.0, 10.0], 'neck_search_bounds': [-0.3, 0.1], 'downsampling_factor_affine': 2}, 'segment': {'downsampling_targets': [2.0, 1.0], 'bias_kernel_width': 70, 'background_mask_sigma': 4.0, 'background_mask_threshold': 0.001, 'mesh_stiffness': 0.1, 'diagonal_covariances': False, 'csf_factor': 0.3}, 'surfaces': {'processes': 2, 'surf': ['lh', 'rh'], 'pial': ['lh', 'rh'], 'vdist': [1.0, 0.75], 'voxsize_pbt': [0.5, 0.25], 'voxsize_refinecs': [0.75, 0.5], 'th_initial': 0.714, 'no_selfintersections': True, 'fillin_gm_from_surf': True, 'open_sulci_from_surf': True, 'exclusion_tissues_fillin_gm': {'left_cerebral_wm': 2, 'right_cerebral_wm': 41, 'stuff_to_exclude': [4, 10, 14, 16, 24, 28, 43, 49, 60]}, 'exclusion_tissues_open_csf': [17, 18, 53, 54]}, 'mesh': {'elem_sizes': {'standard': {'range': [1, 5], 'slope': 1.0}, '1': {'range': [1, 7], 'slope': 1.0}, '2': {'range': [1, 2], 'slope': 1.0}, '5': {'range': [1, 10], 'slope': 0.6}}, 'smooth_size_field': 2, 'skin_facet_size': 2.0, 'facet_distances': {'standard': {'range': [0.1, 3], 'slope': 0.5}}, 'optimize': False, 'remove_spikes': True, 'skin_tag': 1005, 'hierarchy': False, 'smooth_steps': 10, 'skin_care': 20}}
INFO: Using 8 threads, instead of all available.
INFO: Using charm_atlas_mni as charm atlas.
INFO: Reading settings from /Users/sihanliang/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/simnibs/segmentation/atlases/charm_atlas_mni/charm_atlas_mni.ini
INFO: Starting affine registration and neck correction.
INFO: Template registration summary.
INFO: Number of Iterations: 10, Cost: -0.423450

INFO: Adjusting neck.
INFO: Initial cost: -0.05151180803246542
INFO: Neck adjustment done.
INFO: Estimating parameters.
INFO: Starting segmentation.
INFO: ##----------------------------------------------
INFO:               Samsegment Options
INFO: ##----------------------------------------------
INFO: output directory:/Users/sihanliang/Monkeylab/MRI_data/bayi_mri/m2m_bayi/segmentation
INFO: input images: ['bayi_T1w.nii.gz']
INFO: transformed template:/Users/sihanliang/Monkeylab/MRI_data/bayi_mri/m2m_bayi/segmentation/template_coregistered.mgz
INFO: modelSpecifications:{'FreeSurferLabels': [517, 8, 16, 15, 46, 3, 7, 42, 2, 41, 24, 44, 54, 53, 18, 5, 17, 60, 31, 85, 28, 63, 14, 12, 51, 62, 26, 58, 30, 52, 43, 49, 10, 11, 50, 4, 13, 77, 72, 80, 29, 47, 511, 507, 501, 502, 515, 514, 512, 516, 509, 508, 506, 520, 530], 'atlasFileName': '/Users/sihanliang/Monkeylab/MRI_data/bayi_mri/m2m_bayi/segmentation/atlas_level2.txt.gz', 'names': ['Background', 'Left-Cerebellum-Cortex', 'Brain-Stem', '4th-Ventricle', 'Right-Cerebellum-White-Matter', 'Left-Cerebral-Cortex', 'Left-Cerebellum-White-Matter', 'Right-Cerebral-Cortex', 'Left-Cerebral-White-Matter', 'Right-Cerebral-White-Matter', 'Sub-CSF', 'Right-Inf-Lat-Vent', 'Right-Amygdala', 'Right-Hippocampus', 'Left-Amygdala', 'Left-Inf-Lat-Vent', 'Left-Hippocampus', 'Right-VentralDC', 'Left-choroid-plexus', 'Optic-Chiasm', 'Left-VentralDC', 'Right-choroid-plexus', '3rd-Ventricle', 'Left-Putamen', 'Right-Putamen', 'Right-vessel', 'Left-Accumbens-area', 'Right-Accumbens-area', 'Left-vessel', 'Right-Pallidum', 'Right-Lateral-Ventricle', 'Right-Thalamus-Proper', 'Left-Thalamus-Proper', 'Left-Caudate', 'Right-Caudate', 'Left-Lateral-Ventricle', 'Left-Pallidum', 'WM-hypointensities', '5th-Ventricle', 'non-WM-hypointensities', 'Left-undetermined', 'Right-Cerebellum-Cortex', 'Skin', 'Other-Tissues', 'Air-Internal', 'Artery', 'Bone-Cortical', 'Vein', 'Spinal-Cord', 'Bone-Cancellous', 'Mucosa', 'Rectus-Muscles', 'Eyes', 'CorticalCSF', 'Optic-Nerve'], 'colors': [[0, 168, 255, 255], [230, 148, 34, 255], [119, 159, 176, 255], [42, 204, 164, 255], [220, 248, 164, 255], [205, 62, 78, 255], [220, 248, 164, 255], [205, 62, 78, 255], [245, 245, 245, 255], [0, 225, 0, 255], [60, 60, 60, 255], [196, 58, 250, 255], [103, 255, 255, 255], [220, 216, 20, 255], [103, 255, 255, 255], [196, 58, 250, 255], [220, 216, 20, 255], [165, 42, 42, 255], [0, 200, 200, 255], [234, 169, 30, 255], [165, 42, 42, 255], [0, 200, 221, 255], [204, 182, 142, 255], [236, 13, 176, 255], [236, 13, 176, 255], [160, 32, 240, 255], [255, 165, 0, 255], [255, 165, 0, 255], [160, 32, 240, 255], [13, 48, 255, 255], [120, 18, 134, 255], [0, 118, 14, 255], [0, 118, 14, 255], [122, 186, 220, 255], [122, 186, 220, 255], [120, 18, 134, 255], [12, 48, 255, 255], [200, 70, 255, 255], [120, 190, 150, 255], [164, 108, 226, 255], [135, 206, 235, 255], [230, 148, 34, 255], [190, 156, 120, 255], [250, 250, 51, 255], [0, 0, 0, 0], [204, 0, 0, 255], [238, 215, 150, 255], [0, 84, 255, 255], [0, 187, 169, 255], [237, 199, 154, 255], [255, 160, 188, 255], [190, 50, 73, 255], [230, 189, 66, 255], [120, 133, 217, 255], [0, 191, 122, 255]], 'sharedGMMParameters': [GMMparameter(mergedName='Air_Background', numberOfComponents=1, searchStrings=['Background', 'Air-Internal', 'CorticalCSF', 'Bone-Cortical']), GMMparameter(mergedName='Global_WM', numberOfComponents=2, searchStrings=['White', 'Brain-Stem', 'VentralDC', 'Optic-Chiasm', 'Thalamus', 'Optic-Nerve']), GMMparameter(mergedName='Global_GM', numberOfComponents=1, searchStrings=['Cortex', 'Caudate', 'Hippocampus', 'Amygdala', 'Accumbens', 'hypointensities', 'Thalamus']), GMMparameter(mergedName='Global_CSF_messy', numberOfComponents=1, searchStrings=['Sub-CSF', 'vessel', 'choroid-plexus']), GMMparameter(mergedName='Global_CSF_clean', numberOfComponents=1, searchStrings=['Ventricle', 'Inf-Lat-Vent', 'Eyes']), GMMparameter(mergedName='Veins', numberOfComponents=1, searchStrings=['Vein']), GMMparameter(mergedName='Soft_Tissues', numberOfComponents=2, searchStrings=['Other-Tissues', 'Rectus-Muscles', 'Mucosa', 'Skin', 'Artery', 'undetermined', 'Optic-Nerve', 'Bone-Cancellous']), GMMparameter(mergedName='Pallidum', numberOfComponents=1, searchStrings=['Pallidum']), GMMparameter(mergedName='Putamen', numberOfComponents=1, searchStrings=['Putamen']), GMMparameter(mergedName='Spine', numberOfComponents=1, searchStrings=['Spinal-Cord'])], 'useDiagonalCovarianceMatrices': False, 'brainMaskingSmoothingSigma': 4.0, 'brainMaskingThreshold': 0.001, 'K': 0.1, 'biasFieldSmoothingKernelSize': 70}
INFO: optimizationOptions:{'maximumNumberOfDeformationIterations': 20, 'absoluteCostPerVoxelDecreaseStopCriterion': 0.0001, 'verbose': False, 'maximalDeformationStopCriterion': 0.001, 'lineSearchMaximalDeformationIntervalStopCriterion': 0.001, 'maximalDeformationAppliedStopCriterion': 0.0, 'BFGSMaximumMemoryLength': 12, 'multiResolutionSpecification': [{'atlasFileName': '/Users/sihanliang/Monkeylab/MRI_data/bayi_mri/m2m_bayi/segmentation/atlas_level1.txt.gz', 'targetDownsampledVoxelSpacing': 2.0, 'maximumNumberOfIterations': 100, 'estimateBiasField': True}, {'atlasFileName': '/Users/sihanliang/Monkeylab/MRI_data/bayi_mri/m2m_bayi/segmentation/atlas_level2.txt.gz', 'targetDownsampledVoxelSpacing': 1.0, 'maximumNumberOfIterations': 100, 'estimateBiasField': True}]}
INFO: ====================
INFO: {
    depth: 1
    maximumNumberOfDeformationIterations: 20
    absoluteCostPerVoxelDecreaseStopCriterion: 0.0001
    verbose: False
    maximalDeformationStopCriterion: 0.001
    lineSearchMaximalDeformationIntervalStopCriterion: 0.001
    maximalDeformationAppliedStopCriterion: 0.0
    BFGSMaximumMemoryLength: 12
    multiResolutionSpecification: {
        depth: 2
        atlasFileName: /Users/sihanliang/Monkeylab/MRI_data/bayi_mri/m2m_bayi/segmentation/atlas_level1.txt.gz
        targetDownsampledVoxelSpacing: 2.0
        maximumNumberOfIterations: 100
        estimateBiasField: True
    }, {
        depth: 2
        atlasFileName: /Users/sihanliang/Monkeylab/MRI_data/bayi_mri/m2m_bayi/segmentation/atlas_level2.txt.gz
        targetDownsampledVoxelSpacing: 1.0
        maximumNumberOfIterations: 100
        estimateBiasField: True
    }
}
INFO: ====================
INFO: multiResolutionLevel=0
INFO: maximumNumberOfIterations: 100
INFO: Setting up downsampled model
INFO: iterationNumber=0
INFO: EMIterationNumber=0
DEBUG: Traceback
Traceback (most recent call last):
  File "/Users/sihanliang/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/simnibs/cli/charm.py", line 174, in 
    main()
  File "/Users/sihanliang/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/simnibs/cli/charm.py", line 167, in main
    charm_main.run(subject_dir, args.T1, args.T2, args.registerT2, args.initatlas,
  File "/Users/sihanliang/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/simnibs/segmentation/charm_main.py", line 222, in run
    segment_parameters_and_inputs = charm_utils._estimate_parameters(
  File "/Users/sihanliang/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/simnibs/segmentation/charm_utils.py", line 223, in _estimate_parameters
    samsegment.process()
  File "/Users/sihanliang/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/simnibs/segmentation/samseg/Samseg.py", line 154, in process
    self.estimateModelParameters()
  File "/Users/sihanliang/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/simnibs/segmentation/samseg/Samseg.py", line 426, in estimateModelParameters
    downSampledGaussianPosteriors, minLogLikelihood = self.gmm.getGaussianPosteriors(downSampledData,
  File "/Users/sihanliang/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/simnibs/segmentation/samseg/GMM.py", line 137, in getGaussianPosteriors
    gaussianLikelihoods = self.getGaussianLikelihoods(data, mean, variance)
  File "/Users/sihanliang/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/simnibs/segmentation/samseg/GMM.py", line 115, in getGaussianLikelihoods
    L = np.linalg.cholesky(variance)
  File "<__array_function__ internals>", line 5, in cholesky
  File "/Users/sihanliang/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/numpy/linalg/linalg.py", line 763, in cholesky
    r = gufunc(a, signature=signature, extobj=extobj)
  File "/Users/sihanliang/Applications/SimNIBS-4.0/simnibs_env/lib/python3.9/site-packages/numpy/linalg/linalg.py", line 91, in _raise_linalgerror_nonposdef
    raise LinAlgError("Matrix is not positive definite")
numpy.linalg.LinAlgError: Matrix is not positive definite
CRITICAL: Uncaught exception
numpy.linalg.LinAlgError: Matrix is not positive definite

**Comment by liangsihan** [comment]
BTW, I am using Non-human Primate MRI data, could this be the cause of the problem?

**Reply by oulap** [reply]
charm only works with human head scans, so I would expect NHP data to throw an error.
